cmd/backup-manager/app/util/util.go

```diff
@@ -395,6 +400,313 @@ func GetBRMetaData(ctx context.Context, provider v1alpha1.StorageProvider) (*kvb
+}
+
+// initialSnapshotSize calculate size of an initial snapshot in bytes by listing its blocks.
+func initialSnapshotSize(snapshotId string) (uint64, error) {
+	var numBlocks uint64
+	ebsSession, err := NewEBSSession(CloudAPIConcurrency)
+	if err != nil {
+		klog.Errorf("new a ebs session failure.")
+		return 0, err
+	}
+	resp, err := ebsSession.ebs.ListSnapshotBlocks(&ebs.ListSnapshotBlocksInput{
+		SnapshotId: aws.String(snapshotId),
+		MaxResults: aws.Int64(10000),
+	})
+
+	if err != nil {
+		return 0, err
+	}
+
+	blockSize := uint64(*resp.BlockSize)
+	for {
+		numBlocks += uint64(len(resp.Blocks))
+		// check if there is more to retrieve
+		if resp.NextToken == nil {
+			break
+		}
+		resp, err = ebsSession.ebs.ListSnapshotBlocks(&ebs.ListSnapshotBlocksInput{
+			SnapshotId: aws.String(snapshotId),
+			MaxResults: aws.Int64(10000),
+			NextToken:  resp.NextToken,
+		})
+
+		if err != nil {
+			return 0, err
+		}
+	}
+	klog.Infof("full backup snapshot num block %d, block size %d", numBlocks, blockSize)
+	return numBlocks * blockSize, nil
+}
+
+func getPrevSnapshotId(snapshotId string, volSnapshots []*ec2.Snapshot) (string, error) {
+	// sort snapshots by timestamp
+	sort.Slice(volSnapshots, func(i, j int) bool {
+		return volSnapshots[i].StartTime.Before(*volSnapshots[j].StartTime)
+	})
+	var prevSnapshotId string
+	for i, snapshot := range volSnapshots {
+		klog.Infof("the snapshot %s", *snapshot.SnapshotId)
+		if snapshotId == *snapshot.SnapshotId {
+			// first snapshot
+			if i == 0 {
+				return "", nil
+			}
+			prevSnapshotId = *volSnapshots[i-1].SnapshotId
+			break
+		}
+	}
+	if len(prevSnapshotId) == 0 {
+		return "", fmt.Errorf("Could not find the prevousely snapshot id, current snapshotId: %s.", snapshotId)
+	}
+	return prevSnapshotId, nil
+}
+
+// changedBlocksSize calculates changed blocks total size in bytes between two snapshots with common ancestry.
+func changedBlocksSize(preSnapshotId string, snapshotId string) (uint64, error) {
+	var numBlocks uint64
+	ebsSession, err := NewEBSSession(CloudAPIConcurrency)
+	if err != nil {
+		klog.Errorf("new a ebs session failure.")
+		return 0, err
+	}
+	resp, err := ebsSession.ebs.ListChangedBlocks(&ebs.ListChangedBlocksInput{
+		FirstSnapshotId:  aws.String(preSnapshotId),
+		MaxResults:       aws.Int64(10000),
+		SecondSnapshotId: aws.String(snapshotId),
+	})
+
+	if err != nil {
+		return 0, err
+	}
+
+	blockSize := uint64(*resp.BlockSize)
+	klog.Infof("the preSnapshotId %s, the current snapshotId %s", preSnapshotId, snapshotId)
+	for {
+		// retrieve only changed block and blocks only existed in current snapshot (new add blocks)
+		for _, block := range resp.ChangedBlocks {
+			if block.SecondBlockToken != nil && len(aws.StringValue(block.SecondBlockToken)) != 0 {
+				numBlocks += 1
+			}
+		}
+
+		klog.Infof("the current num blocks %d", numBlocks)
+		// check if there is more to retrieve
+		if resp.NextToken == nil || len(aws.StringValue(resp.NextToken)) == 0 {
```

Comment:
this judgement is different from the judgement of `initialSnapshotSize`, do we need the latter condition?

Reply:
fixed in https://github.com/pingcap/tidb-operator/pull/4819/commits/fc18e21a421dee08cf3257099b503047e5d552a7